File: leetcode/201-300/296-smallest-rectangle-enclosing-black-pixels/bench/blackpixels.py
"""LeetCode 296 benchmark lane — Python mirror of blackpixels.kara.

SCALED DOWN to 120 queries against the compiled lanes' 1200. The row/column
probes are per-pixel Python loops over a 4096-wide frame, which is exactly the
shape CPython is worst at, so at full scale this lane would dominate the
suite's wall clock for no extra information. BENCHMARKS.md's rule is to scale
the Python lane rather than let that happen.

The sink therefore does NOT match the other four — it is over a different
number of queries, by construction, and is only comparable to itself.

Uses `bytearray` and plain index loops rather than numpy: the point is to
mirror the same algorithm the other four run, not to measure a vectorised
library that would be doing something structurally different.
"""

N = 4096
QUERIES = 120

img = bytearray(N * N)
W = H = N


def row_has_black(r):
    base = r * W
    for c in range(W):
        if img[base + c] == 1:
            return True
    return False


def col_has_black(c):
    for r in range(H):
        if img[r * W + c] == 1:
            return True
    return False


def _bisect(lo, hi, pred):
    while lo < hi:
        m = lo + (hi - lo) // 2
        if pred(m):
            hi = m
        else:
            lo = m + 1
    return lo


def min_area(x, y):
    top = _bisect(0, x + 1, row_has_black)
    bottom = _bisect(x + 1, H, lambda r: not row_has_black(r))
    left = _bisect(0, y + 1, col_has_black)
    right = _bisect(y + 1, W, lambda c: not col_has_black(c))
    return (bottom - top) * (right - left)


def main():
    r0 = c0 = N // 2
    for r in range(40):
        for c in range(40):
            img[(r0 + r) * N + (c0 + c)] = 1
    for k in range(25):
        img[(r0 + 40 + k) * N + (c0 + 20)] = 1

    checksum = 0
    for q in range(QUERIES):
        sx = r0 + q % 40
        sy = c0 + (q * 7) % 40
        checksum = (checksum * 31 + min_area(sx, sy)) % 1000000007
    print(f"queries {QUERIES} checksum {checksum}")


if __name__ == "__main__":
    main()

File: leetcode/301-400/302-smallest-rectangle-enclosing-black-pixels/bench/blackpixels.rs
// LeetCode 302 benchmark lane — Rust mirror of blackpixels.kara.
//
// Same algorithm: 4-way binary search on the row/column projections. Written as
// plain index loops rather than iterator combinators so the shape matches the
// Kara and C mirrors — an `any()` chain would be measuring a different thing.

struct Image { data: Vec<u8>, w: i64, h: i64 }

impl Image {
    #[inline]
    fn at(&self, r: i64, c: i64) -> u8 { self.data[(r * self.w + c) as usize] }

    fn row_has_black(&self, r: i64) -> bool {
        let mut c = 0;
        while c < self.w { if self.at(r, c) == 1 { return true; } c += 1; }
        false
    }
    fn col_has_black(&self, c: i64) -> bool {
        let mut r = 0;
        while r < self.h { if self.at(r, c) == 1 { return true; } r += 1; }
        false
    }

    fn first_black_row(&self, mut lo: i64, mut hi: i64) -> i64 {
        while lo < hi { let m = lo + (hi - lo) / 2; if self.row_has_black(m) { hi = m } else { lo = m + 1 } }
        lo
    }
    fn first_white_row(&self, mut lo: i64, mut hi: i64) -> i64 {
        while lo < hi { let m = lo + (hi - lo) / 2; if self.row_has_black(m) { lo = m + 1 } else { hi = m } }
        lo
    }
    fn first_black_col(&self, mut lo: i64, mut hi: i64) -> i64 {
        while lo < hi { let m = lo + (hi - lo) / 2; if self.col_has_black(m) { hi = m } else { lo = m + 1 } }
        lo
    }
    fn first_white_col(&self, mut lo: i64, mut hi: i64) -> i64 {
        while lo < hi { let m = lo + (hi - lo) / 2; if self.col_has_black(m) { lo = m + 1 } else { hi = m } }
        lo
    }

    fn min_area(&self, x: i64, y: i64) -> i64 {
        let top = self.first_black_row(0, x + 1);
        let bottom = self.first_white_row(x + 1, self.h);
        let left = self.first_black_col(0, y + 1);
        let right = self.first_white_col(y + 1, self.w);
        (bottom - top) * (right - left)
    }
}

fn main() {
    let n: i64 = 4096;
    let queries: i64 = 1200;
    let mut data = vec![0u8; (n * n) as usize];

    let (r0, c0) = (n / 2, n / 2);
    for r in 0..40 { for c in 0..40 { data[((r0 + r) * n + (c0 + c)) as usize] = 1; } }
    for k in 0..25 { data[((r0 + 40 + k) * n + (c0 + 20)) as usize] = 1; }
    let img = Image { data, w: n, h: n };

    let mut checksum: i64 = 0;
    for q in 0..queries {
        let sx = r0 + q % 40;
        let sy = c0 + (q * 7) % 40;
        checksum = (checksum * 31 + img.min_area(sx, sy)) % 1000000007;
    }
    println!("queries {} checksum {}", queries, checksum);
}
